lib/Onnxifi/ThreadPool.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "ThreadPool.h"
+
+#include <iostream>
+
+namespace glow {
+namespace onnxifi {
+
+ThreadPool::ThreadPool(unsigned numWorkers) : shouldStop_(false) {
+  // Intialize all workers and make each one run threadPoolWorkerMain.
+  for (unsigned i = 0; i < numWorkers; i++) {
+    std::thread th(std::bind(&ThreadPool::threadPoolWorkerMain, this));
+    workers_.push_back(std::move(th));
+  }
+}
+
+ThreadPool::~ThreadPool() {
+  // Signal to workers to stop.
+  shouldStop_ = true;
+
+  // Notify all worker threads in case any are waiting on the condition
+  // variable.
+  cond_.notify_all();
+
+  // Join all worker threads.
+  for (auto &w : workers_) {
+    w.join();
+  }
+  workers_.clear();
+}
+
+void ThreadPool::submit(const std::function<void(void)> &fn) {
+  // Add fn to the work queue.
+  std::lock_guard<std::mutex> lock(mtx_);
+  workQueue_.push(fn);
+  cond_.notify_one();
+}
+
+void ThreadPool::threadPoolWorkerMain() {
+  std::unique_lock<std::mutex> lock(mtx_, std::defer_lock);
+
+  while (!shouldStop_) {
+    // Lock the lock after processing a work item.
+    lock.lock();
+
+    // If work queue is empty, wait to be signalled when
+    // a work item is submitted.
+    while (workQueue_.empty() && !shouldStop_) {
+      cond_.wait(lock);
+    }
+
+    // If shouldStop_ was set to false while the thread
+    // was asleep, break out of the main loop.
+    if (shouldStop_) {
+      break;
+    }
+
+    // Pop a work item from the queue, and make sure to unlock
+    // the lock before processing it.
+    auto workItem = workQueue_.front();
+    workQueue_.pop();
+    lock.unlock();
+
+    // Process work item.
+    try {
+      workItem();
+    } catch (const std::bad_function_call &e) {
```

Comment:
Let's leave off the try/catch for now and discuss `-fno-exceptions` in a separate PR.  Sorry for the churn on this.